spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/mapping/SpannerObjectMapperImpl.java

```diff
@@ -0,0 +1,183 @@
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Predicate;
+
+import com.google.cloud.spanner.Mutation.WriteBuilder;
+import com.google.cloud.spanner.ResultSet;
+import com.google.cloud.spanner.Struct;
+import com.google.cloud.spanner.Type;
+import com.google.cloud.spanner.ValueBinder;
+
+import org.springframework.beans.BeanUtils;
+import org.springframework.data.mapping.PersistentPropertyAccessor;
+import org.springframework.data.mapping.PropertyHandler;
+import org.springframework.util.Assert;
+
+/**
+ * @author Chengyuan Zhao
+ */
+public class SpannerObjectMapperImpl implements SpannerObjectMapper {
+
+	private final SpannerMappingContext spannerMappingContext;
+
+	public SpannerObjectMapperImpl(SpannerMappingContext spannerMappingContext) {
+		Assert.notNull(spannerMappingContext,
+				"A valid mapping context for Spanner is required.");
+		this.spannerMappingContext = spannerMappingContext;
+	}
+
+	@Override
+	public <T> List<T> mapToUnmodifiableList(ResultSet resultSet, Class<T> entityClass) {
+		ArrayList<T> result = new ArrayList<>();
+		while (resultSet.next()) {
+			result.add(read(entityClass, resultSet.getCurrentRowAsStruct()));
+		}
+		return Collections.unmodifiableList(result);
+	}
+
+	@Override
+	public <R> R read(Class<R> type, Struct source) {
+		R object = BeanUtils.instantiateClass(type);
+		SpannerPersistentEntity<?> persistentEntity = this.spannerMappingContext
+				.getPersistentEntity(type);
+		PersistentPropertyAccessor accessor = persistentEntity
+				.getPropertyAccessor(object);
+
+		for (Type.StructField field : source.getType().getStructFields()) {
+			String name = field.getName();
+			SpannerPersistentProperty property = persistentEntity
+					.getPersistentPropertyByColumnName(name);
+			if (source.isNull(name)) {
+				continue;
+			}
+
+			boolean valueSet = false;
+
+			for (Method method : source.getClass().getMethods()) {
+				// the retrieval methods are named like getDate or getTimestamp
+				if (!method.getName().startsWith("get")) {
+					continue;
+				}
+
+				Class[] params = method.getParameterTypes();
+				if (params.length != 1 || !name.getClass().isAssignableFrom(params[0])) {
+					continue;
+				}
+
+				Class retType = method.getReturnType();
+				if (retType.isAssignableFrom(property.getType())) {
+					try {
+						accessor.setProperty(property, method.invoke(source, name));
+						valueSet = true;
+						break;
+					}
+					catch (IllegalAccessException e) {
+						throw new UnsupportedOperationException(
+								"Could not set value for read.", e);
+					}
+					catch (InvocationTargetException e) {
+						throw new UnsupportedOperationException(
+								"Could not set value for read.", e);
+					}
+				}
+			}
+
+			if (!valueSet) {
+				throw new UnsupportedOperationException(
+						"No compatible types in object for reading column: " + name);
```

Comment:
nit: "object" is confusing. I'd say: "None of the fields of class " + type + " can be used for reading column " + name